app/lib/oidc_client.rb

```diff
@@ -0,0 +1,131 @@
+require "openid_connect"
+
+class OidcClient
+  class OAuthFailure < RuntimeError; end
+
+  attr_reader :client_id,
+              :provider_uri
+
+  delegate :authorization_endpoint,
+           :token_endpoint,
+           :userinfo_endpoint,
+           :end_session_endpoint,
+           to: :discover
+
+  def initialize(provider_uri, client_id, secret)
+    @provider_uri = provider_uri
+    @client_id = client_id
+    @secret = secret
+  end
+
+  def auth_uri
+    nonce = SecureRandom.hex(16)
+    {
+      uri: client.authorization_uri(
+        scope: scopes,
+        state: nonce,
+        nonce: nonce,
+      ),
+      state: nonce,
+    }
+  end
+
+  def redirect_uri
+    host =
+      if ENV["VCAP_APPLICATION"].present?
+        "https://#{JSON.parse(ENV['VCAP_APPLICATION'])['uris'].first}"
+      else
+        "http://#{ENV['VIRTUAL_HOST']}"
+      end
```

Comment:
How does this fit into our normal app runtime environment?  Is the `VIRTUAL_HOST` env var set in AWS?

Reply:
Looks like `VIRTUAL_HOST` is something we set in the `docker-compose.yml` for `finder-frontend` in govuk-docker, but not in any of the AWS environments.  I stole this code from an app deployed to the PaaS (barking permit, I think), so I'll update this to use something we have available in the AWS environment.

Reply:
I think `GOVUK_WEBSITE_ROOT` is the env var I'm looking for.

Reply:
:+1: I don't think we need `VCAP_APPLICATION` as that was just useful for testing things on the PaaS, but we're now getting closer to running it for real.  So `ENV["VIRTUAL_HOST"] || ENV["GOVUK_WEBSITE_ROOT"]` is probably what we want.